### PR TITLE
feat(cb2-9601): update vehicle configuration options

### PR DESCRIPTION
**Pull request by owen-corrigan-bjss**
CB2-9601

THE WORK FOR THIS TICKET IS NOT COMPLETE. This PR is to give everyone access to the resolver that null's tech record fields based on enums. 

The functionality for the ticket is complete but there are tests to write for the resolver and the enum needs swapping for the types package. 

_One line description_
[CB2-9601](https://dvsa.atlassian.net/browse/CB2-9601)

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts [hunk 1]
```diff
@@ -205,7 +205,8 @@ export class TechRecordSummaryChangesComponent implements OnInit, OnDestroy {
   }
 
   isNotEmpty(value: unknown): boolean {
-    if (value == null || value === '') return false;
+    if (value === null) return true;
```
Comment (naathanbrown) [inline src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:208]: Would this not just fall into the return true at the bottom?

File: src/app/resolvers/tech-record-review/tech-record-review.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import { selectTechRecord, updateEditingTechRecord } from '@store/technical-records';
+import {
+  catchError,
+  map,
+  of,
+  take,
+  tap,
+} from 'rxjs';
+
+export const TechRecordReviewResolver: ResolveFn<boolean> = () => {
```
Comment (naathanbrown) [inline src/app/resolvers/tech-record-review/tech-record-review.resolver.ts:19]: This is a function, so naming convention needs to be techRecordReviewResolver

File: src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts [hunk 3]
```diff
@@ -205,8 +205,8 @@ export class TechRecordSummaryChangesComponent implements OnInit, OnDestroy {
   }
 
   isNotEmpty(value: unknown): boolean {
-    if (value == null || value === '') return false;
-    if (typeof value === 'object') return Object.values(value).length > 0;
+    if (value === '') return false;
+    if (typeof value === 'object' && value !== null) return Object.values(value).length > 0;
```
Comment (owen-corrigan-bjss) [inline src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: had to check that value is not null here otherwise it won't drop through to the final return

**Review by shivangidas — APPROVED** [review]

File: src/app/features/tech-record/tech-record-routing.module.ts [hunk 4]
```diff
@@ -40,7 +40,10 @@ const routes: Routes = [
       isCustomLayout: true,
     },
     canActivate: [MsalGuard, RoleGuard],
-    resolve: { techRecord: techRecordViewResolver },
+    resolve: {
+      techRecord: techRecordViewResolver,
+      load: techRecordReviewResolver,
```
Comment (shivangidas) [inline src/app/features/tech-record/tech-record-routing.module.ts:45]: the resolver name could be changed maybe 😁

File: src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts (hunk 3 above)
Comment (pbardy2000) [reply src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: Just want to double check if this is an issue or not, this checks for exact equality to null, and thus will allow undefined - is that the behaviour you want?

Comment (owen-corrigan-bjss) [reply src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: @pbardy2000 So we will only be wanting to let null through as the resolver is nulling the fields that contain bad data rather than setting them to undefined. I'll extend the empty string check to check for undefined

Comment (Twebb97) [reply src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: Is there a test for this that has an explicit null value ?

File: src/app/resolvers/tech-record-validate/tech-record-validate.resolver.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if (record[key as keyof TechRecordVehicleType<'hgv'>]) {
+      const valid = Object.values(key).includes(checks[`${key}`]);
+      if (!valid) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (validatedRecord as any)[`${key}`] = null;
+      }
+    }
+  });
+  return validatedRecord as TechRecordType<'put'>;
+};
```
Comment (pbardy2000) [inline src/app/resolvers/tech-record-validate/tech-record-validate.resolver.ts:112]:
```suggestion
const handlePsv = (record: TechRecordVehicleType<'psv'>) => {
  const validatedRecord: TechRecordVehicleType<'psv'> = { ...record };

  const checks = {
    techRecord_vehicleConfiguration: HgvPsvVehicleConfiguration,
  } as const;

  type Checks = typeof checks;
  type CheckKey = keyof Checks;

  const keys = Object.keys(checks) as CheckKey[];

  keys.forEach((key) => {
    if (record[key as keyof TechRecordVehicleType<'psv'>]) {
      const valid = Object.values(checks).includes(checks[key]);
      if (!valid) {
        validatedRecord[key] = null;
      }
    }
  });
  return validatedRecord as TechRecordType<'put'>;
};

const handleTrl = (record: TechRecordVehicleType<'trl'>) => {
  const validatedRecord: TechRecordVehicleType<'trl'> = { ...record };

  const checks = {
    techRecord_vehicleConfiguration: TrlVehicleConfiguration,
  };

  type Checks = typeof checks;
  type CheckKey = keyof Checks;

  const keys = Object.keys(checks) as CheckKey[];

  keys.forEach((key: CheckKey) => {
    if (record[key as keyof TechRecordVehicleType<'trl'>]) {
      const valid = Object.values(checks).includes(checks[key]);
      if (!valid) {
        validatedRecord[key] = null;
      }
    }
  });
  return validatedRecord as TechRecordType<'put'>;
};

const handleHgv = (record: TechRecordVehicleType<'hgv'>) => {
  const validatedRecord: TechRecordVehicleType<'hgv'> = { ...record };

  const checks = {
    techRecord_vehicleConfiguration: HgvPsvVehicleConfiguration,
  };

  type Checks = typeof checks;
  type CheckKey = keyof Checks;

  const keys = Object.keys(checks) as CheckKey[];

  keys.forEach((key: CheckKey) => {
    if (record[key as keyof TechRecordVehicleType<'hgv'>]) {
      const valid = Object.values(checks).includes(checks[key]);
      if (!valid) {
        validatedRecord[key] = null;
      }
    }
  });
  return validatedRecord as TechRecordType<'put'>;
};
```

Do you not want something more like this?

File: src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts (hunk 3 above)
Comment (Twebb97) [reply src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: Ive got 2 things here:

1. lets add a test for an object with null to be allowed througj

2. We're now saying that we allow null but not empty in a function called isNotEmpty which would raise some questions in the future if I saw this again. In the object world this would break the O in solid, even more so that it was discussed that this would move into a more utils type file in the future.

Hence is there a more appropriate name and can we self document the code, or do we default to a comment ?

File: src/app/forms/custom-sections/body/body.component.ts [hunk 6]
```diff
@@ -104,9 +105,18 @@ export class BodyComponent implements OnInit, OnChanges, OnDestroy {
   }
 
   get bodyTypes(): MultiOptions {
-    const optionsMap = vehicleBodyTypeCodeMap.get(this.techRecord.techRecord_vehicleType as VehicleTypes);
-    const values = [...optionsMap!.values()];
+    let vehicleType: string = this.techRecord.techRecord_vehicleType;
+
+    if (this.techRecord.techRecord_vehicleType === 'hgv' && this.techRecord.techRecord_vehicleConfiguration === 'articulated') {
```
Comment (Twebb97) [inline src/app/forms/custom-sections/body/body.component.ts:110]: Are we going to see more of this type of change in the future for other specific vehicle types ?
if so is there going to be a massive if else chain that can be avoided early ?

File: src/app/forms/validators/custom-validators.ts [hunk 7]
```diff
@@ -298,4 +298,17 @@ export class CustomValidators {
       return null;
     };
   };
+
+  static updateFunctionCode = (): ValidatorFn => {
+    return (control: AbstractControl): ValidationErrors | null => {
+      const vehicleFunctionCode = control.root.get('techRecord_functionCode');
+      if (control.dirty) {
+        if (control?.value === 'rigid') vehicleFunctionCode?.setValue('R', { emitEvent: false });
```
Comment (Twebb97) [inline src/app/forms/validators/custom-validators.ts:306]: You can avoid this using a map with types and thier function codes and someting like this 

vehicleFunctionCode?.setValue(functionCode.get(TYPE_VARIBALE), { emitEvent: false })

File: src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts (hunk 3 above)
Comment (owen-corrigan-bjss) [reply src/app/features/tech-record/components/tech-record-summary-changes/tech-record-summary-changes.component.ts:209]: Both fair points Tom I'll work on those

**Review by shivangidas — APPROVED** [review]

File: src/app/forms/custom-sections/body/body.component.ts (hunk 6 above)
Comment (shivangidas) [reply src/app/forms/custom-sections/body/body.component.ts:110]: looking at the ticket, it appears all vehicles have different conditions now for displaying the  https://dvsa.atlassian.net/browse/CB2-9601
